### PR TITLE
Bringing custom shellcontextmenu

### PR DESCRIPTION
**Pull request by BuraChuhadar**

### CODE REVIEW

**Comment by yaichenbaum** [comment]
@BuraChuhadar This is really exciting! Do you want me to mark this a a draft for now?

**Comment by BuraChuhadar** [comment]
Sounds good please mark it as draft. I am still working on it.

**Comment by yaichenbaum** [comment]
@BuraChuhadar Will this solve https://github.com/duke7553/files-uwp/issues/719?

**Comment by BuraChuhadar** [comment]
@yair100 I am currently working on phase 1. So I am not planning to bring File > New options yet. For those the full menu needs to re-written to close #719 . I am not sure how do you want to approach to that.

**Comment by yair100** [comment]
> @yair100 I am currently working on phase 1. So I am not planning to bring File > New options yet. For those the full menu needs to re-written to close #719 . I am not sure how do you want to approach to that.

Lets keep this limited to open with for now, however to plan ahead, is there a central place in the registry to pull a list of the item types you can create?

**Comment by lampenlampen** [comment]
I think i need some clarification on this.

As far as i understand this retrives the "open with" options for a filetype from the registry.
Does this include additional options, than using Windows ["ApplicationPicker"](https://docs.microsoft.com/en-us/windows/uwp/launch-resume/launch-the-default-app-for-a-file#launch-the-file)? Or what are the benefits of using this approach instead of the "ApplicationPicker".

**Comment by BuraChuhadar** [comment]
You cannot bring File > New options from the registry. You would need to read it from the shell 32 DLL instead.

File: Files/UserControls/LayoutModes/PhotoAlbum.xaml [hunk 1]
```diff
@@ -271,14 +271,14 @@
...
-        Background="{StaticResource ApplicationPageBackgroundThemeBrush}"
         ContextFlyout="{StaticResource BaseLayoutContextFlyout}">
         <ProgressBar
             Name="ProgBar"
             Height="10"
             VerticalAlignment="Top"
             IsIndeterminate="True"
-            Visibility="{x:Bind AssociatedViewModel.LoadIndicator.IsVisible, Mode=TwoWay, UpdateSourceTrigger=PropertyChanged}" />
+            Visibility="{x:Bind AssociatedViewModel.LoadIndicator.IsVisible, Mode=TwoWay, UpdateSourceTrigger=PropertyChanged}"
+            Background="Transparent" />
```
Comment (duke7553) [inline Files/UserControls/LayoutModes/PhotoAlbum.xaml:281]: Does setting a transparent background here fix anything? Just curious...

File: Files/Helpers/RegistryReader.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Files.Helpers
+{
+    class RegistryReader
+    {
+        private bool IsAdministrator()
+        {
+            return (new WindowsPrincipal(WindowsIdentity.GetCurrent()))
```
Comment (duke7553) [inline Files/Helpers/RegistryReader.cs:15]: Are Universal Windows apps currently allowed to access WindowsPrincipal?

**Review by lukeblevins — CHANGES_REQUESTED** [review]
Please address comment regarding WindowsPrincipal

**Comment by duke7553** [comment]
> Sounds good please mark it as draft. I am still working on it.

Just saw this now. Let me know if this should be a draft. 

Hopefully my feedback helps guide you, but I really like these changes so far,

File: Files/UserControls/LayoutModes/PhotoAlbum.xaml (hunk 1 above)
Comment (BuraChuhadar) [reply Files/UserControls/LayoutModes/PhotoAlbum.xaml]: This is not my changes. You can ignore that line.

File: Files/Helpers/RegistryReader.cs (hunk 2 above)
Comment (BuraChuhadar) [reply Files/Helpers/RegistryReader.cs:15]: I am seeing it as supported under UWP: https://stackoverflow.com/questions/36713824/how-to-get-group-infolike-guest-admin-etc-of-current-login-user-in-uwp-windo. Let me know if this is a concern.

Comment (BuraChuhadar) [reply Files/Helpers/RegistryReader.cs:15]: I am seeing it as supported under UWP: https://stackoverflow.com/questions/36713824/how-to-get-group-infolike-guest-admin-etc-of-current-login-user-in-uwp-windo. Let me know if this is a concern.

**Comment by BuraChuhadar** [comment]
@yair100 I finalized the code no need to draft it. Please take a look. I am seeing WindowsPrincipal class is being supported. If you still think that isn't the case or if there is a way to check if it is supported under UWP then I can take a look.

**Review by BuraChuhadar — COMMENTED** [review]
I am seeing it as supported under UWP: https://stackoverflow.com/questions/36713824/how-to-get-group-infolike-guest-admin-etc-of-current-login-user-in-uwp-windo. Let me know if this is a concern.

File: Files/Helpers/RegistryReader.cs (hunk 2 above)
Comment (BuraChuhadar) [reply Files/Helpers/RegistryReader.cs:15]: I am seeing it as supported under UWP: https://stackoverflow.com/questions/36713824/how-to-get-group-infolike-guest-admin-etc-of-current-login-user-in-uwp-windo. Let me know if this is a concern.

**Comment by duke7553** [comment]
@BuraChuhadar The stack overflow link you sent stated that the API is intended for use inside a brokered component rather than the UWP component. 

It doesn't seem supported here: https://docs.microsoft.com/en-us/dotnet/api/?view=dotnet-uwp-10.0&term=principal

**Comment by BuraChuhadar** [comment]
OK, I will take a look no problem to address WIndows Principal code.

**Comment by duke7553** [comment]
@BuraChuhadar Actually, I think it would be better if we removed that part altogether because modern Windows will essentially deprecate the concept of admin vs standard user. With the WinRT sandbox, users expect to be standard users, so we should always assume they are.

**Comment by BuraChuhadar** [comment]
@duke7553 it has been a while I couldn't focus on this. I added the changes you mentioned. Let's finish phase one of this implementation. I will continue adding more features around this. Let me know if it looks good now. The administration security check has been removed from the code.

**Review by lukeblevins — COMMENTED** [review]
Looks very good so far!

**Comment by BuraChuhadar** [comment]
@duke7553 can you please merge this?

**Comment by yaichenbaum** [comment]
@BuraChuhadar I tested the pr and I didn't see any visual changes.

**Comment by BuraChuhadar** [comment]
@yair100 it looks in these areas for menu options:

CurrentUser\\Software\\Classes\\{fileExtension}\\shell
CurrentUser\\Software\\Classes\\*\\shell

If there is none then you won't see anything. What is the expected behavior on your end?

**Comment by tsvietOK** [comment]
@BuraChuhadar Seems like your patch broke "Open in terminal" option

**Comment by BuraChuhadar** [comment]
@tsvietOK fixed. @duke7553 any other feedback ready to merge?

**Comment by yaichenbaum** [comment]
I don't see any visual change with this pull request.

**Comment by xPoppyx** [comment]
@yaichenbaum maybe it's for the "open with" option?

**Comment by yaichenbaum** [comment]
@BuraChuhadar What options is this meant to add to the context menu?

**Comment by lukeblevins** [comment]
@yair100 This PR searches a registry key for certain options that should be in the context menu, and adds them. 
@BuraChuhadar I will test this today, and hopefully we can merge it for the next release

**Comment by yaichenbaum** [comment]
@duke7553 I tested this, and even though I have multiple items in the menu that show in File Explorer, I don't see them when running the code from this pull request.

**Comment by lukeblevins** [comment]
@yair100 The items this PR adds to the context flyout should be specific to the file extension, so if you install 7-Zip, I'd assume the option for it would display for compressed archives. It won't show everything either, because some items require admin access which is not expected in the UWP app model.

**Review by lukeblevins — CHANGES_REQUESTED** [review]
@BuraChuhadar When you get a chance, I would recommend adding back the part that reads from HKEY_CLASSES_ROOT. To the best of my knowledge, we won't need the admin check because we're only reading from it.

**Comment by BuraChuhadar** [comment]
Just an update here. I didn't have the time to integrate the requests yet. I will be working on this once I get the chance.

**Comment by yaichenbaum** [comment]
@BuraChuhadar Thank you for the update, we are looking forward to merging this when ready.

**Comment by BuraChuhadar** [comment]
@yair100 the item you requested has been added to the code. I also made some major changes code wise around the Files.Launcher. Let me know your thoughts.

**Comment by MikeCodesDotNET** [comment]
Love this PR! I'm using it as the basis of a little hack I'm playing with myself. 

Only issue is I've found a small bug. Skype seems to not provide a name, thus an empty menu item appears in the context list.

empty_context_item

## Problem 
context-bug

## Quick Fix
I fixed this in my local branch by using adding a LINQ  statement on the return value to select only values with a commandname. 

`                
return shellList.Where(x => !string.IsNullOrEmpty(x.commandName));
`

**Comment by BuraChuhadar** [comment]
@MikeCodesDotNET will check once I get the chance thanks.

**Comment by BuraChuhadar** [comment]
OK new features:

1- Added MUIVerb Support. This will fix "Share with Skype" issue.
2- Directory registry parsing was fixed. 

Screenshot:

Annotation 2020-06-11 233219

**Comment by BuraChuhadar** [comment]
Next update I will try to bring the icons which are associated with the context menus.

**Comment by yaichenbaum** [comment]
@BuraChuhadar Is the PR ready for review?

**Comment by BuraChuhadar** [comment]
Let me add the icon integration then let's review please.

**Comment by BuraChuhadar** [comment]
@yair100 this is ready for review. I took a step back and changed my decision from adding the icons due to restrictions of IconElement class under MenuFlyout class. I would wait to see WinUI 3.0 will deliver something better. 

I am going to attach this discussion related to the FlyoutMenu icons: https://github.com/microsoft/microsoft-ui-xaml/issues/1494

I updated the current:

1- Bringing CommandLineToArgvW to parse command registry from Win32API
2- Execution of the context menu was a problem for some applications. This has been fixed.
3- Improved MUIVerb support
4- Changed the ordering of the context menu items. Screenshot:

image

At this point this PR getting quite large. I would suggest let's merge this in as I will continue with other approaches to improve the overall quality with the context menus to pair up the functionality as much as I can with Windows' native explorer.

I am looking forward to hearing about the feedback.

**Review by lukeblevins — APPROVED** [review]
Very excellent work! Users will love this feature

**Comment by tsvietOK** [comment]
Exception when right clicking on the txt file.
image

**Comment by BuraChuhadar** [comment]
@tsvietOK I will add a null check around that in that case you won't be able to load the extra shell menu items. Thank you for sharing it.

**Comment by BuraChuhadar** [comment]
Never mind. I found the problem. I sent a fix for it and fixed the merge conflicts.

**Comment by tsvietOK** [comment]
@BuraChuhadar For some reason extended context menu displaying only after second right click on folder. As i understand for the moment extended context menu is not working for files.

**Comment by BuraChuhadar** [comment]
@tsvietOK it should work for files as well. I didn't try directly right clicking. I will check once I have the chance.

**Comment by BuraChuhadar** [comment]
@tsvietOK sent another fix. I see that the opening of the context menu wasn't synchronizing with the UI correctly given that the registry call method was async. I sent a fix by simply waiting for the registry method read task to complete. Please let me know if you see anything else. Thanks.

**Comment by yaichenbaum** [comment]
@BuraChuhadar I tested the latest changes and for the first time, I was able to see it working 🎉. It feels kind of off without the icons, we are using custom icons in menus in a couple of places so if you are running into issues there let us know so we can work together to improve the UI.

**Comment by tsvietOK** [comment]
@BuraChuhadar I am probably doing something wrong, but when i right clicking on file, displaying only standard context menu. Can you share example screenshot on how it should look like.

**Comment by BuraChuhadar** [comment]
@yair100 I won't be addressing the icons as I mentioned above. Native explorer has the same behavior. You are welcome to send another PR for the icons.

@tsvietOK please send me your reproducible steps. Otherwise, I would suggest merging these changes and address issues on later tickets. You are welcome to assign to me the upcoming issues.

Here is the video of working changes with directory and files. It is a youtube video click on it to see it under youtube:

[](http://www.youtube.com/watch?v=30XEUf_vIzg "Files UWP Demo")

**Comment by tsvietOK** [comment]
@BuraChuhadar maybe i don't have supported apps installed.
Can you check, should Notepad++ be displayed when right clicking on file?

**Comment by BuraChuhadar** [comment]
@tsvietOK this feature won't use shell context extensions. Which in this case Notepad++ is using shell context extensions.  I am seeing there are articles around Microsoft not interested in supporting shell context extensions and yet there is still software out there using this. Even PowerToys under Microsoft project is using this.

I will try to bring shell context extension support with a different PR at the moment this PR supports only static shell context menus.

**Comment by duke7553** [comment]
@BuraChuhadar If this is ready to merge, I'd like to do so considering the vast amount of work you put into this. I'm really impressed!

**Comment by BuraChuhadar** [comment]
@duke7553 it is ready to merge.

File: Files/Views/LayoutModes/GridViewBrowser.xaml [hunk 3]
```diff
@@ -154,8 +154,8 @@
             </MenuFlyoutItem>
         </MenuFlyout>
 
-        <MenuFlyout x:Key="BaseLayoutItemContextFlyout" Opening="RightClickItemContextMenu_Opening">
-            <MenuFlyoutItem
+            <MenuFlyout x:Key="BaseLayoutItemContextFlyout" x:Name="BaseLayoutItemContextFlyout" Opening="RightClickContextMenu_Opening">
```
Comment (gave92) [inline Files/Views/LayoutModes/GridViewBrowser.xaml:157]: Is this correct? Shouldn't this be:
```cs
<MenuFlyout 
    x:Key="BaseLayoutItemContextFlyout" 
    x:Name="BaseLayoutItemContextFlyout" 
    ----> Opening="RightClick***Item***ContextMenu_Opening">
```

**Comment by XPoppyX** [comment]
> when i right clicking on file, displaying only standard context menu.

the same happens to me

**Comment by yair100** [comment]
> @yair100 I won't be addressing the icons as I mentioned above. Native explorer has the same behavior. You are welcome to send another PR for the icons.

@BuraChuhadar Great work on this PR! As far as I can tell, File Explorer does show the icons in the right click context menu.

File: Files/Views/LayoutModes/GridViewBrowser.xaml (hunk 3 above)
Comment (BuraChuhadar) [reply Files/Views/LayoutModes/GridViewBrowser.xaml:157]: No, that one is correct.

**Comment by BuraChuhadar** [comment]
@XPoppyX you won't see all the available items in context menu from the explorer. This feature only implements the static context menus, not the dynamic ones. File a bug and assign it to me if you see there are potential issues with reproducible steps.

**Comment by BuraChuhadar** [comment]
@yair100 not all the items under the native explorer has icons that's what I meant. MenuFlyout also doesn't support bitmap images so I am OK leaving like this. If you have a proposal or a helper method to assign bitmaps to a menu flyout component then I can bring the icons.

**Comment by XPoppyX** [comment]
@BuraChuhadar #1172

**Comment by yair100** [comment]
> @yair100 not all the items under the native explorer has icons that's what I meant. MenuFlyout also doesn't support bitmap images so I am OK leaving like this. If you have a proposal or a helper method to assign bitmaps to a menu flyout component then I can bring the icons.

That makes sense, in the future we should try to get that working. In the meantime your work helped implement one of the most requested features 🎉.

**Comment by BuraChuhadar** [comment]
@yair100 implementing that would be challenging because the dynamic shell context menus are using COM implementation and depending on the STA or MTA thread apartment the rendering of the shell context menu changes. I am thinking to see if I can bring something by using PowerShell or cmd to query the shell context menu and render it like that. I will create another PR once I see it is possible.

**Comment by XPoppyX** [comment]
@BuraChuhadar #1185

File: Files/Views/LayoutModes/GridViewBrowser.xaml (hunk 3 above)
Comment (BuraChuhadar) [reply Files/Views/LayoutModes/GridViewBrowser.xaml:157]: My apologies you are right.  https://github.com/files-community/files-uwp/issues/1185 will address this.

File: Files/Views/LayoutModes/GenericFileBrowser.xaml.cs [hunk 4]
```diff
@@ -104,12 +104,12 @@ public override void SetSelectedItemOnUi(ListedItem item)
...
 
-        public override void SetSelectedItemsOnUi(List<ListedItem> items)
+        public override void SetSelectedItemsOnUi(List<ListedItem> selectedItems)
         {
             ClearSelection();
-            foreach (ListedItem item in items)
+            foreach (ListedItem selectedItem in selectedItems)
             {
-                AllView.SelectedItems.Add(item);
+                AllView.SelectedItem = selectedItem;
```
Comment (tsvietOK) [inline Files/Views/LayoutModes/GenericFileBrowser.xaml.cs:112]: What purpose of this change? This broke file selection. File size is not more visible when changing view mode from `GridViewBrowser` to `GenericFileBrowser` and back.

**Pull request merged**